src/shared/vscode/commands2.ts

```diff
@@ -0,0 +1,330 @@
+    }
+
+    public get registered() {
+        return !!this.subscription
+    }
+
+    public build(...args: Parameters<T>): Builder {
+        const id = this.resource.id
+
+        return {
+            asUri: this.buildUri(id, args),
+            asCommand: this.buildCommand(id, args),
+            asCodeLens: this.buildCodeLens(id, args),
+            asTreeNode: this.buildTreeNode(id, args),
+        }
+    }
+
+    public register(...args: U): RegisteredCommand<T> {
+        const { id, name } = this.resource
+        const label = name ? `"${name}" (id: ${id})` : `"${id}"`
+        const target = this.resource.factory(...args)
+        const instrumented = (...args: Parameters<T>) => runCommand(target, { label, args })
+        this.subscription = this.commands.registerCommand(this.resource.id, instrumented)
+
+        return this
+    }
+
+    public async execute(...args: Parameters<T>): Promise<ReturnType<T> | undefined> {
+        return this.commands.executeCommand<ReturnType<T>>(this.resource.id, ...args)
+    }
+
+    public dispose(): void {
+        this.disposed = true
+        this.subscription?.dispose()
+    }
+
+    private buildUri(id: string, args: unknown[]) {
+        return () => vscode.Uri.parse(`command:${id}?${encodeURIComponent(JSON.stringify(args))}`)
+    }
+
+    private buildCommand(id: string, args: unknown[]) {
+        return (content: PartialCommand) => ({ ...content, command: id, arguments: args })
+    }
+
+    private buildCodeLens(id: string, args: unknown[]) {
+        return (range: vscode.Range, content: PartialCommand) => {
+            return new vscode.CodeLens(range, this.buildCommand(id, args)(content))
+        }
+    }
+
+    private buildTreeNode(id: string, args: unknown[]) {
+        return (content: PartialTreeItem) => {
+            return new (class extends AWSTreeNodeBase {
+                public constructor() {
+                    super(content.label, vscode.TreeItemCollapsibleState.None)
+                    this.id = `${id}-${(CommandResource.counter += 1)}`
+                    this.command = { command: id, arguments: args, title: content.label }
+
+                    Object.assign(this, content)
+                }
+            })()
+        }
+    }
+}
+
+interface CommandInfo<T extends Callback> {
+    readonly label: string
+    readonly args: Parameters<T>
+    readonly logging?: boolean
+    readonly errorHandler?: (error: Error) => ReturnType<T> | void
+}
+
+async function runCommand<T extends Callback>(fn: T, info: CommandInfo<T>): Promise<ReturnType<T> | undefined> {
+    const { args, label, logging } = { logging: true, ...info }
+    const logger = logging ? getLogger() : new NullLogger()
+    const withArgs = args.length > 0 ? ` with arguments [${args.map(String).join(', ')}]` : ''
+    logger.debug(`command: running ${label}${withArgs}`)
+
+    try {
+        const result = await fn(...args)
+        logger.debug(`command: ${label} succeeded`)
```

Comment:
this is where we'd add generic telemetry? maybe worth a TODO comment

Reply:
Yup you'd add it right there.